frameless-runtime/src/utxo.rs

```diff
@@ -0,0 +1,253 @@
+    pub witness: Sig,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Output {
+    /// The address that owns this output. Based on either a public key or a Tuxedo Piece
+    pub redeemer: Redeemer,
+    /// The data associated with this output. In the simplest case, this will be a token balance, but could be arbitrarily rich state.
+    pub data: Value,
+    /// An Id for this type Such that we know how to encode or decode the 'data' field
+    pub data_id: TypeId,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Transaction {
+    /// The inputs refer to currently existing unspent outputs that will be consumed by this transaction
+    pub inputs: Vec<Input>,
+    /// Similar to inputs, Peeks refer to currently existing utxos, but they will be read only, and not consumed
+    pub peeks: Option<Vec<Input>>,
+    /// The new outputs to be created by this transaction.
+    pub outputs: Vec<Output>,
+}
+
+pub type Utxo = Output;
+pub type UtxoRef = OutputRef;
+
+pub trait Redeem {
+    fn redeem(self, tx: &[u8], witness: &[u8]) -> bool;
+}
+
+impl Redeem for Redeemer {
+    fn redeem(self, tx: &[u8], witness: &[u8]) -> bool {
+        let signature = match Signature::try_from(&witness[..]) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+        sp_io::crypto::sr25519_verify(&signature, &tx, &Public::from_h256(self))
+    }
+}
+
+// TODO: Implement this for Each Tuxedo Piece
+pub trait UtxoSet {
+
+    /// TODO: Change these bool return types to Result types for more error propagation clarity
+
+    /// Check whether a given utxo exists in the current set
+    fn contains(utxo_ref: UtxoRef) -> bool;
+
+    /// Insert the given utxo into the state storing it with the given ref
+    /// The ref is probably the hash of a tx that created it and its index in that tx, but this decision is opaque to this trait
+    /// Return whether the operation is successful (It can fail if the ref is already present)
+    fn insert(utxo_ref: UtxoRef, utxo: Utxo) -> bool;
+
+    ///
+    /// nullify the utxo by either:
+    /// - Consuming it entirely
+    /// - Putting it on "Timeout"
+    /// - Not consuming it but marking it as spent
+    ///
+    fn nullify(utxo_ref: UtxoRef) -> Option<Utxo>;
+
+    fn peak(utxo_ref: UtxoRef) -> Option<Utxo> {
+        let encoded_utxo = sp_io::storage::get(&utxo_ref.encode())?;
+        match Utxo::decode(&mut &encoded_utxo[..]) {
+            Ok(utxo) => Some(utxo),
+            Err(_) => None,
+        }
+    }
+}
+
+pub trait Get<T> {
+    fn get(&self) -> T;
+}
```

Comment:
There are a few things, like this for example, that are currently in `frame_support`, but are useful outside of frame. Another one I encountered is `ensure!`. Once we have a sense of what these things are, we should PR Substrate to move them to a less frame-specific place. Maybe it warrants a new crate called something like `runtime_support`.

Reply:
I did notice there are several little things in frame that are really not tied down to frame at all.